### ISSUE TITLE
Create floating index for About page

### ISSUE THREAD

**Issue by vladh** [opensourcepledge/opensourcepledge.com#200]
Our [About page][about] is growing and is likely to grow further. As it grows, it's becoming a little bit confusing to keep track of where you are on the page, and it's not currently possible to see an overview. We should create something like a floating table of contents in a sidebar so that users can keep track of the page's contents. On mobile, the table of contents could move to the top, perhaps collapsed by default.

[about]: https://opensourcepledge.com/about

**Comment by safwanyp** [comment]
I would love to contribute for this. I have a very very quick draft of something. Could use some early feedback if possible!

The text looks too big for the index, but I haven't added any custom css for that. It's inferred from the root. Do you think it should become smaller?

![Screenshot 2024-10-14 at 22 30 49](https://github.com/user-attachments/assets/56276398-14bc-4170-b924-6c5733d5fff9)

**Comment by vladh** [comment]
Hey @safwanyp, thank you for this mockup, and thanks for being willing to help! :)

Here are some points from me:

* There isn't enough differentiation between the index items, they kind of blend into each other. There should be a bit more vertical space between the index items, maybe `0.3rem`.
* I think things would be clearer if each index item (and associated heading) had a number, so `1.`, `2.` and so on.
* The index box has more vertical padding than horizontal padding — I think they should be equal.
* The top of the index box should never be higher than the top of the first heading. This shows that the index box is not within the scope of the whole website, or the global navigation, but rather is visually within the scope of this specific page.
* It's not fully clear to me how the index box is positioned. Is it anchored to the side of the viewport, or to the page's content box? That is, if I made the viewport wider, does the index box move, or not? I think the index box should not move, so I think it should be anchored to the side of the content box.
* I think we need to rethink how this page is written altogether a little bit. In your implementation, when a heading is added, does an index item need to be manually added as well? We should avoid this, because it will get tedious. There should be some kind of system where either (1) the index is automatically generated from the headings, or (2) both the index and the page's content are automatically generated from some other source, such as a folder of Markdown files.
* Of course, for any implementation we're going to need to decide how it works for every screen size. It would be great to see how this implementation translates to small mobile devices. I think a collapsed-by-default approach could work — see the mobile navigation on [vladh.net](https://vladh.net) for example.

**Comment by safwanyp** [comment]
Appreciate the feedback! I work on this only for like an hour every night after work, so apologies for the slow responses!

1. I added a `0.2rem` spacing along with incremental numbering to the list as you mentioned.
2. Fixed the padding.
3. The `TableOfContents.astro` component is nothing more than a div with an `<ol>` (and a button for smaller sreens). The items are populated by the page with a tiny script which I plan to extract somehow for re-usability. I don't think it's perfect, but this was what I could come up with. If you have suggestions, let me know!
```js
// this script generates the Table of Contents elements from the sections in the page
  document.addEventListener('DOMContentLoaded', () => {
    const sections = extractSections();
    // extractSections is a function to create an array of objects from the sections in the page.
    const tocElement = document.querySelector('div.toc ol');
    if (tocElement) {
      tocElement.innerHTML = sections.map(section =>
        `<li><a href="#${section.id}">${section.title}</a></li>`
      ).join('');
    }
  });
```
4. For smaller screens I added a button to the top, similar to what's on your site. 
<img width="1724" alt="Screenshot 2024-10-15 at 22 00 14" src="https://github.com/user-attachments/assets/984182c9-bb97-46ef-83cf-5c4d79272f12">

I don't have a clean implementation of the correct positioning and stuff of the component, but I hope to have one soon (hopefully tomorrow). The main problem I'm facing is the component affecting the bounding box of the content on the page itself. I _think_ I will either have to change the `Layout` to account for this new component or figure out a way to do it without touching the `Layout`.
